builtin/mainmenu/init.lua

```diff
@@ -47,9 +47,49 @@ local tabs = {
 }
 
 --------------------------------------------------------------------------------
+local function show_exit_dialog(tabview, show_dialog)
+	tabview:hide()
+	local dlg = dialog_create(
+		"mainmenu_quit_confirm",
+		function()
+			return "size[10,3,true]" ..
+				"label[0.5,0.5;" .. fgettext("Are you sure you want to quit?") .. "]" ..
+				"checkbox[0.5,1;cb_show_dialog;" .. fgettext("Always show this dialog.") .. ";" .. tostring(show_dialog) .. "]" ..
+				"style[btn_quit_confirm_yes;bgcolor=red]" ..
+				"button[0.5,2.0;2.5,0.5;btn_quit_confirm_yes;" .. fgettext("Quit") .. "]" ..
+				"button[7.0,2.0;2.5,0.5;btn_quit_confirm_cancel;" .. fgettext("Cancel") .. "]"
+		end,
+		function(this, fields)
+			if fields.cb_show_dialog ~= nil then
+				core.settings:set_bool("enable_esc_dialog", core.is_yes(fields.cb_show_dialog))
+				return false
+			elseif fields.btn_quit_confirm_yes then
+				this:delete()
+				core.close()
+				return true
+			elseif fields.btn_quit_confirm_cancel or fields.key_escape or fields.quit then
+				this:delete()
+				if tabview and tabview.show then
+					tabview:show()
+				end
+				return true
+			end
+		end,
+		nil
```

Comment:
Cross-reference to https://github.com/luanti-org/luanti/pull/16390#issuecomment-3221090424